<a href="https://colab.research.google.com/github/SerArtDev/redes-neuronales/blob/main/redes_neuronales_keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Primer red neuronal con Keras
Entre las distintas bibliotecas para deep learning, Keras es la más fácil y amigable sin que esto signifique que está limitada. Con Keras se puede realizar redes enomes y complejas.

Keras trabaja sobre TensorFlow, otra biblioteca popular en este ámbito, de tal forma que es necesario tener esta segunda biblioteca.

## Modelo de Regresión

En este caso vamos a generar una red neuronal para los datos de muestras de cemento en los cuales tienen distintas cantidades de componentes para variar su fuerza de compresión.

| Cement | Blast furnace Slag | Fly ash | Water | Superplasticizer | Coarse Aggregate | Fine Aggregate | Age | Strength |
|--------|--------------------|---------|-------|------------------|------------------|----------------|-----|----------|
| 540.0  | 0.0                | 0.0     | 162.0 | 2.5              | 1040.0           | 676.0          | 28  | 79.99    |
| 540.0  | 0.0                | 0.0     | 162.9 | 2.5              | 1055.0           | 676.0          | 28  | 61.89    |
| 332.5  | 142.5              | 0.0     | 228.0 | 0.0              | 932.0            | 594.0          | 365 | 41.05    |
| 332.5  | 142.5              | 0.0     | 228.0 | 0.0              | 932.9            | 594.0          | 365 | 41.05    |
| 198.6  | 132.4              | 0.0     | 192.0 | 0.0              | 978.4            | 825.5          | 360 | 44.30    |


Queremos determinar la fuerza (Strength) basado en los otros cinco parámetros. Para esto creamos una red neuronal con 8 datos de entrada y uno de salida. La capa de entrada tiene 8 neuronas, mientras que la de salida tiene 1 neurona. También cuenta con dos capas ocultas, cada una con 5 neuronas. Todas las neuronas están conectadas con todas las neuronas de la capa anterior; a esto se le llama una red densa (dense network).


In [ ]:
import pandas as pd

concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
predictors = concrete_data.drop('Strength', axis=1)
target = concrete_data['Strength']

El modelo de nuestra red consiste en una secuencia lineal de capas, por lo cual usaremos el modelo secuencial. Cada capa se añade con el método ´add´, que recibe como entrada la capa con sus parámetros. Para la primer capa oculta es necesario pasar el número de entradas a la red (8 para este caso).

In [ ]:
import keras
from keras.models import Sequential
from keras.layers import Dense

n_inputs = predictors.shape[1]

model = Sequential()
model.add(Dense(5, activation='relu', input_shape=(n_inputs,)))
model.add(Dense(5, activation='relu'))
model.add(Dense(1))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Teniendo definido el modelo, definimos parámetros de entrenamiento, como el método de optimización y la función de costo o error.

In [ ]:
model.compile(optimizer='adam', loss='mean_squared_error')
model.fit(predictors, target, epochs=50)

Epoch 1/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - loss: 14884.5986
Epoch 2/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 2913.8506
Epoch 3/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 1082.8527
Epoch 4/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 777.0303
Epoch 5/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 762.6771
Epoch 6/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 662.6004
Epoch 7/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 687.7938
Epoch 8/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 656.6331
Epoch 9/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 619.7487
Epoch 10/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 571.8511
Epoch 11/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 533.2031
Epoch 12/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 502.7973
Epoch 13/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 431.0143
Epoch 14/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 463.8586
Epoch 15/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - 

Teniendo el modelo entrenado, se pueden hacer predicciones usando el método predict.

In [ ]:
import numpy as np
test_value = np.array([540, 0, 0, 162, 2.5, 1040, 676, 28])
prediction = model.predict(test_value.reshape(1, -1))
print("Predicted strength: %.2f" % prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 63ms/step
Predicted strength: 39.68


<ipython-input-4-f0a82528643c>:4: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  print("Predicted strength: %.2f" % prediction)


## Modelo de Clasificación

Para este caso, se tienen datos de hongos clasificados en comestibles y venenosos junto a sus características. La idea es hacer una red neuronal que clasifique a los hondos en una de estas dos clases.


In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("uciml/mushroom-classification")

print("Path to dataset files:", path)

100%|██████████| 34.2k/34.2k [00:00<00:00, 8.38MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/uciml/mushroom-classification/versions/1


In [ ]:
data = pd.read_csv(path + "/mushrooms.csv")
data.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

Las redes neuronales funcionan con datos numéricos, por lo tanto, estos datos categóricos de esta base de datos deben ser convertidos. El método get_dummies de pandas permite esto creando una columna por cada categoría y asignando un valor false o true.

In [ ]:
target = data['class']
predictors = data.drop('class', axis=1)
target = pd.get_dummies(target)
predictors = pd.get_dummies(predictors)
target.head()


e      p
0  False   True
1   True  False
2   True  False
3  False   True
4   True  False

In [ ]:
predictors.head()

cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0        False        False        False        False        False   
1        False        False        False        False        False   
2         True        False        False        False        False   
3        False        False        False        False        False   
4        False        False        False        False        False   

   cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  cap-surface_y  \
0         True          False          False           True          False   
1         True          False          False           True          False   
2        False          False          False           True          False   
3         True          False          False          False           True   
4         True          False          False           True          False   

   ...  population_s  population_v  population_y  habitat_d  habitat_g  \
0  ...          True         False         False      False      False   
1  ...         False         False         False      False       True   
2  ...         False         False         False      False      False   
3  ...          True         False         False      False      False   
4  ...         False         False         False      False       True   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0      False      False      False       True      False  
1      False      False      False      False      False  
2      False       True      False      False      False  
3      False      False      False       True      False  
4      False      False      False      False      False  

[5 rows x 117 columns]

La capa de entrada de nuestra red debe tener la cantidad total de clases de los predictores. El método get_dummies generó 117 columnas, así que esa es la cantidad de neuronas de entrada necesarias. Para la capa de salida, se debe poner la cantidad de clases de salida, en este caso, venenoso y comestible.
Se elige de forma arbitraria dos capas ocultas con 15 neuronas cada una. La capa de salida utiliza la función de activación softmax, que esta es útil para problemas de clasificación. Sin embargo, la capa oculta usa la función de activación ReLU.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
classificator = Sequential()
classificator.add(Dense(15, activation='relu', input_shape=(predictors.shape[1],)))
classificator.add(Dense(15, activation='relu'))
classificator.add(Dense(2, activation='softmax'))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Se puede usar una fracción de los datos para entrenar el modelo y el resto para comprobar si funciona correctamente usando sklearn.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)


Ahora podemos entrenar el modelo usando una función de costo específica para problemas de clasificación.

In [ ]:
classificator.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
classificator.fit(X_train, y_train, epochs=20)

Epoch 1/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8169 - loss: 0.4637
Epoch 2/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9958 - loss: 0.0287
Epoch 3/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9988 - loss: 0.0069
Epoch 4/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 0.0027
Epoch 5/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 1.0000 - loss: 0.0014
Epoch 6/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 1.0000 - loss: 6.7511e-04
Epoch 7/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 1.0000 - loss: 3.7612e-04
Epoch 8/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 1.0000 - loss: 2.4212e-04
Epoch 9/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 1.0000 - loss: 1.8397e-04
Epoch 10/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 1.4928e-04
Epoch 11/20
178/178 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 9.3124e-05
Epoch 12/20
178/178 ━━━━━━━━

Se puede evaluar el modelo para saber si está funcionando correctamente usando los datos con los que no fue entrenado.

In [ ]:
test_loss, test_acc = classificator.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 1.4726e-05
Test accuracy: 1.0
